examples/seq2seq/finetune_trainer.py

```diff
@@ -0,0 +1,397 @@
+            for d in [model.model.encoder, model.model.decoder]:
+                freeze_params(d.embed_positions)
+                freeze_params(d.embed_tokens)
+        except AttributeError:
+            freeze_params(model.shared)
+            for d in [model.encoder, model.decoder]:
+                freeze_params(d.embed_tokens)
+
+    if model_args.freeze_embeds:
+        freeze_embeds(model)
+    if model_args.freeze_encoder:
+        freeze_params(model.get_encoder())
+        assert_all_frozen(model.get_encoder())
+
+    if isinstance(tokenizer, MBartTokenizer) or isinstance(tokenizer, MarianTokenizer):
+        dataset_class = TranslationDataset
+    else:
+        dataset_class = Seq2SeqDataset
+
+    # Get datasets
+    train_dataset = (
+        dataset_class(
+            tokenizer,
+            type_path="train",
+            data_dir=data_args.data_dir,
+            n_obs=data_args.n_train,
+            max_target_length=data_args.max_target_length,
+            max_source_length=data_args.max_source_length,
+            prefix=model.config.prefix or "",
+        )
+        if training_args.do_train
+        else None
+    )
+    eval_dataset = (
+        dataset_class(
+            tokenizer,
+            type_path="val",
+            data_dir=data_args.data_dir,
+            n_obs=data_args.n_val,
+            max_target_length=data_args.val_max_target_length,
+            max_source_length=data_args.max_source_length,
+            prefix=model.config.prefix or "",
+        )
+        if training_args.do_eval
+        else None
+    )
+    test_dataset = (
+        dataset_class(
+            tokenizer,
+            type_path="test",
+            data_dir=data_args.data_dir,
+            n_obs=data_args.n_test,
+            max_target_length=data_args.test_max_target_length,
+            max_source_length=data_args.max_source_length,
+            prefix=model.config.prefix or "",
+        )
+        if training_args.do_predict
+        else None
+    )
+
+    # Initialize our Trainer
+    trainer = Seq2SeqTrainer(
+        model=model,
+        args=training_args,
+        train_dataset=train_dataset,
+        eval_dataset=eval_dataset,
+        data_collator=Seq2SeqDataCollator(tokenizer, data_args),
+        compute_metrics=build_compute_metrics_fn(data_args.task) if training_args.predict_from_generate else None,
+        prediction_loss_only=training_args.prediction_loss_only,
+    )
+
+    # Training
+    if training_args.do_train:
+        trainer.train(
+            model_path=model_args.model_name_or_path if os.path.isdir(model_args.model_name_or_path) else None
+        )
+        trainer.save_model()
+        # For convenience, we also re-save the tokenizer to the same directory,
+        # so that you can share your model easily on huggingface.co/models =)
+        if trainer.is_world_master():
+            tokenizer.save_pretrained(training_args.output_dir)
+
+    # Evaluation
+    eval_results = {}
```

Comment:
Only doing eval here on validation data, thinking test eval should handled by `run_eval.py`

Reply:
why? 
Does this run every epoch?

Reply:
No, this runs after the training is finished when `--do_eval` is set. If `run_eval` can be used for evaluation on test data then why should we duplicate here ?

Reply:
Super annoying to run two commands the first of which takes a long time.
I also don't understand the distinction you are making between val and test.
But if it makes your code way cleaner to ignore test we can just write a wrapper later that a runs reasonable of commands.

Reply:
I will add that it might make the workflow easier to add a bunch of extras/QOL improvements in subsequent PRs, all my comments don't need to addressed here.

Reply:
okay, I'll try to add that as well.

> I also don't understand the distinction you are making between val and test.

This how every other trainer script is written, evaluate on val after training is finished if  `--do_eval` is set and on test of `--do_predict` is set.

Reply:
Note that you can set `--do_train` to False, so you shouldn't have a different evaluation script.